### ISSUE TITLE
[feature] Generate esm-specific declaration files

### ISSUE THREAD

**Issue by IanVS** [egoist/tsup#760]
We are using tsup in storybook to bundle our files and create `.d.ts` files.  A user, @weswigham,  [explained](https://github.com/storybookjs/storybook/pull/19269#discussion_r991455547) to us that we need to actually create separate type definition files for esm and cjs.  Is this something that others have encountered?  If so, maybe it makes sense for tsup to generate both `.d.ts` and `.d.mts` files.

It looks like this is a typescript 4.7+ feature, and is explained in https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#package-json-exports-imports-and-self-referencing, which shows separate declaration files for the esm and cjs exports.

The `.d.mts` and `.d.cts` extensions themselves are supported from 4.5+ (https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#new-file-extensions)

**Comment by cyco130** [comment]
My current workaround is to use the following in the config file (I have the inverse issue):

```ts
async onSuccess() {
  await cp(
    resolve(__dirname, "dist/index.d.ts"),
    resolve(__dirname, "dist/index.d.cts"),
  );
},
```

(Needs more work for more complex libs, of course).

**Comment by aleclarson** [comment]
Surprisingly, this `exports` map doesn't satisfy TypeScript. It seems the `.d.ts` needs to be `.d.mts` or else you get TS2307 errors on import.
```
"./pg/*": {
  "import": "./dist/pg/*/index.mjs",
  "types": "./dist/pg/*/index.d.ts"
},
```
<sub>Hey there, @cyco130  and @IanVS 😄 </sub>

**Comment by RichiCoder1** [comment]
For some extra context: https://github.com/pmndrs/zustand/issues/1381

TL;DR: TypeScript treats the below as CJS even when technically MJS is used under the covers:
```
"export": {
  ".": {
    "types": "dist/index.d.ts",
    "import": "dist/index.mjs",
    "require": "dist/index.js"<
  }
}
```
Or vice verse if the module is `"type": "module"`.

So the missing part of `tsup` is to emit two different declarations in the above case (`["dist/index.d.ts", "dist/index.d.mts"]`) compiled with the correct targe. Further more, the output of the Module (ESM) dts needs to have file extensions in any imports otherwise TypeScript will break when running under `"target": "node16"` or target `"target": "nodenext"`.

**Comment by brc-dd** [comment]
The `onSuccess` thing doesn't work because of #700 👀

**Comment by cyco130** [comment]
> The `onSuccess` thing doesn't work because of #700 eyes

Ooh, I must be copying the old `d.ts` then! Thanks for the heads up.
